components/brave_wallet/browser/meld_integration_response_parser.cc

```diff
@@ -0,0 +1,411 @@
+        service_provider_value->name, service_provider_value->service_provider,
+        service_provider_value->status, service_provider_value->website_url,
+        service_provider_value->categories,
+        ParseOptionalMapOfStrings(service_provider_value->category_statuses),
+        std::move(logos).value_or(nullptr));
+
+    service_providers.emplace_back(std::move(sp));
+  }
+
+  if (service_providers.empty()) {
+    return std::nullopt;
+  }
+
+  return service_providers;
+}
+
+std::optional<std::vector<std::string>> ParseMeldErrorResponse(
+    const base::Value& json_value) {
+  // Parses results like this:
+  // {
+  //     "code": "BAD_REQUEST",
+  //     "message": "Bad request",
+  //     "errors": [
+  //         "[amount] Must be a decimal value greater than zero"
+  //     ],
+  //     "requestId": "eb6aaa76bd7103cf6c5b090610c31913",
+  //     "timestamp": "2022-01-19T20:32:30.784928Z"
+  // }
+
+  const auto meld_error_value =
+      meld_integration_responses::MeldError::FromValue(json_value);
+  if (!meld_error_value) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is not a dict";
+    return std::nullopt;
+  }
+
+  std::vector<std::string> errors;
+  if (meld_error_value->errors && !meld_error_value->errors->empty()) {
+    errors.assign(meld_error_value->errors->begin(),
+                  meld_error_value->errors->end());
+  }
+
+  if (meld_error_value->message.has_value() && errors.empty()) {
+    errors.emplace_back(*meld_error_value->message);
+  }
+
+  if (errors.empty()) {
+    return std::nullopt;
+  }
+
+  return errors;
+}
+
+std::tuple<std::optional<std::vector<mojom::MeldCryptoQuotePtr>>,
+           std::optional<std::string>>
+ParseCryptoQuotes(const base::Value& json_value) {
+  // Parses results like this:
+  // {
+  //   "quotes": [
+  //     {
+  //       "transactionType": "CRYPTO_PURCHASE",
+  //       "sourceAmount": 50,
+  //       "sourceAmountWithoutFees": 43.97,
+  //       "fiatAmountWithoutFees": 43.97,
+  //       "destinationAmountWithoutFees": null,
+  //       "sourceCurrencyCode": "USD",
+  //       "countryCode": "US",
+  //       "totalFee": 6.03,
+  //       "networkFee": 3.53,
+  //       "transactionFee": 2,
+  //       "destinationAmount": 0.00066413,
+  //       "destinationCurrencyCode": "BTC",
+  //       "exchangeRate": 75286,
+  //       "paymentMethodType": "APPLE_PAY",
+  //       "customerScore": 20,
+  //       "serviceProvider": "TRANSAK"
+  //     }
+  //   ],
+  //   "message": null,
+  //   "error": null
+  // }
+  std::optional<std::string> error;
+  const auto quote_resp_value =
+      meld_integration_responses::CryptoQuoteResponse::FromValue(json_value);
+  if (!quote_resp_value) {
+    LOG(ERROR) << "Invalid response, could not parse JSON, JSON is not a dict";
+    return std::tuple<std::optional<std::vector<mojom::MeldCryptoQuotePtr>>,
+                      std::optional<std::string>>{std::nullopt, error};
+  }
+
+  if (quote_resp_value->error) {
```

Comment:
It returs {nullopt,nullopt} - is that expected behavior?

Reply:
Yes it could be in case we can not parse `CryptoQuoteResponse` correctly.